File: revolut/CurrencyManager.m
//
//  CurrencyManager.m
//  revolut
//
//  Created by Alexander Danilov on 02/09/2017.
//  Copyright © 2017 Home. All rights reserved.
//

#import "CurrencyManager.h"
#import "ApiService.h"

@implementation CurrencyManager {
    ApiService *_apiService;
}
    
double const RATE_UPDATE_INTERVAL = 30.0;
    
@synthesize rates;

-(instancetype)initWithApiService:(ApiService *)apiService {
    self = [super init];
    if (self) {
        _apiService = apiService;
    }
    return self;
}
    
-(void)startUpdatingRates {
    [self updateRates];
}
    
-(void)updateRates {
    [_apiService getCurrencyRates:^(NSDictionary *newRates, NSError *error) {
        if (error == nil) {
            [newRates setValue:[[NSDecimalNumber alloc] initWithInt:1] forKey:@"EUR"];
            [self setValue:newRates forKey:@"rates"];
        } else {
            [self onError:error];
        }
        [self performSelector:@selector(updateRates) withObject:nil afterDelay:RATE_UPDATE_INTERVAL]; // delay can be estimated using last request time
    }];
}

-(void (^)())subscribeOnError:(void (^)(NSError *))handler {
    id observer = [[NSNotificationCenter defaultCenter] addObserverForName:@"Error" object:self queue:nil usingBlock:^(NSNotification *notificaion) {
        NSError *error = [notificaion.userInfo valueForKey:@"error"];
        handler(error);
    }];
    
    return ^{
        [[NSNotificationCenter defaultCenter] removeObserver:observer];
    };
}

-(void)onError:(NSError *)error {
    [[NSNotificationCenter defaultCenter] postNotificationName:@"Error" object:self userInfo: @{@"error": error}];
}

-(NSDecimalNumber *)getExchangeRateOfValue:(NSDecimalNumber *)value from:(NSString *)fromCurrency to:(NSString *)toCurrency {
    if (!rates || !rates[fromCurrency] || !rates[toCurrency]) return nil;
    
    NSDecimalNumber *fromRate = rates[fromCurrency];
    NSDecimalNumber *toRate = rates[toCurrency];
    
    return [[value decimalNumberByMultiplyingBy:toRate] decimalNumberByDividingBy:fromRate];
}

@end

File: revolut/CurrencyManager.swift
//
//  CurrencyManager.swift
//  revolut
//
//  Created by Alexander Danilov on 13/08/2018.
//  Copyright © 2018 Home. All rights reserved.
//

import Foundation

class CurrencyManager: NSObject {
    static let RATE_UPDATE_INTERVAL = 30.0
    
    let apiService: ApiService
    
    var rates: NSDictionary?
    
    init(apiService: ApiService) {
        self.apiService = apiService
        
        super.init()
    }
    
    func startUpdatingRates() {
        self.apiService.getCurrencyRates({ newRates, error in
            if let error = error {
                self.onError(error)
                return
            }
            
            if var newRates = newRates {
                newRates["EUR"] = NSDecimalNumber(value: 1)
                self.setValue(newRates, forKey: "rates")
            }
            
            self.perform(#selector(self.startUpdatingRates), with: nil, afterDelay: CurrencyManager.RATE_UPDATE_INTERVAL) // TODO delay can be estimated using last request time
        })
    }

    func subscribeOnError(_ handler: @escaping (Error) -> Void) -> () -> Void {
        let observer = NotificationCenter.default.addObserver(forName: NSNotification.Name("Error"), object: self, queue: nil, using: { notificaion in
            let error = notificaion.userInfo!["error"] as! Error
            handler(error)
        })
        return {
            NotificationCenter.default.removeObserver(observer)
        }
    }

    private func onError(_ error: Error) {
        NotificationCenter.default.post(name: NSNotification.Name("Error"), object: self, userInfo: ["error": error])
    }
    
    func getExchangeRate(ofValue value: NSDecimalNumber, from fromCurrency: String, to toCurrency: String) -> NSDecimalNumber? {
        guard
            let rates = self.rates as? [String: NSDecimalNumber],
            let fromRate = rates[fromCurrency],
            let toRate = rates[toCurrency]
            else { return nil }
        
        return value.multiplying(by: toRate).dividing(by: fromRate)
    }

}
